client/state/sites/domains/actions.js

```diff
@@ -0,0 +1,130 @@
+/**
+ * External dependencies
+ */
+import debugFactory from 'debug';
+import map from 'lodash/map';
+
+/**
+ * Internal dependencies
+ */
+import { createSiteDomainObject } from './assembler';
+import i18n from 'lib/mixins/i18n';
+import wpcom from 'lib/wp';
+import {
+	SITE_DOMAINS_RECEIVE,
+	SITE_DOMAINS_REQUEST,
+	SITE_DOMAINS_REQUEST_SUCCESS,
+	SITE_DOMAINS_REQUEST_FAILURE,
+} from 'state/action-types';
+
+/**
+ * Module vars
+ */
+const debug = debugFactory( 'calypso:site-domains:actions' );
+
+/**
+ * `action-creator`
+ * Returns an action object to be used in signalling that
+ * an object containing the domains for
+ * a given site have been received.
+ *
+ * @param {Number} siteId - identifier of the site
+ * @param {Object} response - WP REST-API domains endpoint response
+ * @returns {Object} the action object
+ */
+export const createDomainsReceiveAction = ( siteId, response ) => {
+	const { domains } = response;
+
+	return {
+		type: SITE_DOMAINS_RECEIVE,
+		siteId,
+		domains: map( domains, createSiteDomainObject )
+	};
+};
+
+/**
+ * `action-creator`
+ * Return SITE_DOMAINS_REQUEST action object
+ *
+ * @param {Number} siteId - side identifier
+ * @return {Object} siteId - action object
+ */
+export const createDomainsRequestAction = siteId => {
+	return {
+		type: SITE_DOMAINS_REQUEST,
+		siteId
+	};
+};
+
+/**
+ * `action-creator`
+ * Return SITE_DOMAINS_REQUEST_SUCCESS action object
+ *
+ * @param {Number} siteId - side identifier
+ * @return {Object} siteId - action object
+ */
+export const createDomainsRequestSuccessAction = siteId => {
+	return {
+		type: SITE_DOMAINS_REQUEST_SUCCESS,
+		siteId
+	};
+};
+
+/**
+ * `action-creator`
+ * Return SITE_DOMAINS_REQUEST_FAILURE action object
+ *
+ * @param {Number} siteId - site identifier
+ * @param {Object} error - error message according to REST-API error response
+ * @return {Object} action object
+ */
+export const createDomainsRequestFailureAction = ( siteId, error ) => {
+	return {
+		type: SITE_DOMAINS_REQUEST_FAILURE,
+		siteId,
+		error: error ? String( error ) : Boolean( error )
+	};
+};
+
+/**
+ * Fetches domains for the given site.
+ *
+ * @param {Number} siteId - identifier of the site
+ * @returns {Function} a promise that will resolve once fetching is completed
+ */
+export function fetchSiteDomains( siteId ) {
+	return dispatch => {
+		dispatch( createDomainsRequestAction( siteId ) );
+
+		return new Promise( resolve => {
+			wpcom
+			.site( siteId )
+			.domainsList( ( error, data ) => {
```

Comment:
Can you check if `wpcom.site( siteId ).domainsList()` returns a promise? Most of the wpcom methods support both callback styles and promise styles, if it does, you can return the wpcom call instead of creating a new promise above.